File: src-tauri/src/domain/aggregates/security.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SecurityType {
    Stock,
    Fund,
    Etf,
    Bond,
    Gold,
    Option,
    Other,
}

impl std::fmt::Display for SecurityType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Stock => write!(f, "stock"),
            Self::Fund => write!(f, "fund"),
            Self::Etf => write!(f, "etf"),
            Self::Bond => write!(f, "bond"),
            Self::Gold => write!(f, "gold"),
            Self::Option => write!(f, "option"),
            Self::Other => write!(f, "other"),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Security {
    pub id: Uuid,
    pub symbol: String,
    pub name: String,
    pub security_type: SecurityType,
    pub exchange: Option<String>,
    pub currency_code: String,
    pub current_price: Option<rust_decimal::Decimal>,
}

impl Security {
    pub fn new(
        id: Uuid,
        symbol: String,
        name: String,
        security_type: SecurityType,
        exchange: Option<String>,
        currency_code: String,
        current_price: Option<rust_decimal::Decimal>,
    ) -> Self {
        Self {
            id,
            symbol,
            name,
            security_type,
            exchange,
            currency_code,
            current_price,
        }
    }
}

File: yucai/server/internal/holding/ent/schema/security.go
package schema

import (
	"entgo.io/ent"
	"entgo.io/ent/dialect/entsql"
	entschema "entgo.io/ent/schema"
	"entgo.io/ent/schema/field"
	"entgo.io/ent/schema/index"
	"github.com/google/uuid"
	"time"
)

// Security is a global reference to a tradable security (no tenant).
type Security struct {
	ent.Schema
}

func (Security) Annotations() []entschema.Annotation {
	return []entschema.Annotation{
		entsql.WithComments(true),
	}
}

func (Security) Fields() []ent.Field {
	return []ent.Field{
		field.UUID("id", uuid.UUID{}).
			Default(uuid.New),
		field.String("symbol").
			NotEmpty(),
		field.String("name").
			NotEmpty(),
		field.String("security_type").
			Comment("stock, fund, etf, bond, gold, option, other"),
		field.String("exchange").
			Optional().
			Default(""),
		field.String("currency_code").
			Default("CNY"),
		field.Int64("current_price_cents").
			Optional().
			Default(0),
		field.Time("created_at").
			Default(time.Now).
			Immutable(),
	}
}

func (Security) Edges() []ent.Edge {
	return nil
}

func (Security) Indexes() []ent.Index {
	return []ent.Index{
		index.Fields("symbol", "exchange").Unique(),
	}
}
